### PR TITLE
feat(archive): allow specifying a compression level

### LINKED ISSUE #2574: Allow setting a compression level for archives
# Summary

TarStrategy should allow setting an optional compression level.
Behavior should be backwards compatible with current workflows and implementation.

# Motivation

When outputs are large files or directories with terrible compression ratios (e.g. build caches, binaries, random data, already compressed or encrypted files) - the extra gzip pass with default compression is terrible for performance and could/should be skipped.

In these cases, `tar + gzip.NoCompression` vs `tar + gzip.DefaultCompression` shows an immediate performance boost with a relatively tiny network overhead traded against giant CPU overhead.

Our major use case was a 3GB+ Bazel build cache directory that showed a difference of ~10sec vs ~2min.

# Proposal

- Add an optional compression level field to TarStrategy.
  - When omitted, it should default to `gzip.DefaultCompression` which is the current behavior.
- Use it directly when archiving files in the `BaseImagePath` case.
- Pass it to the relevant `CopyFile`-capable executors (pns+docker) and implement as needed.
  - Since the docker executor uses the `gzip` command, "round up" `NoCompression` to `--fast`.

---

**Message from the maintainers**:

If you wish to see this enhancement implemented please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

### PR DESCRIPTION
**Pull request by adamglt**
Allows setting a compression level for an archive's TarStrategy.
Default falls back to `gzip.DefaultCompression`, which is the current behavior.

Closes #2574 which contains more info.

### CODE REVIEW

**Comment by adamglt** [comment]
Seems like a flaky test - rerun?

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 1]
```diff
@@ -642,7 +642,11 @@ type ArchiveStrategy struct {
 }
 
 // TarStrategy will tar and gzip the file or directory when saving
-type TarStrategy struct{}
+type TarStrategy struct {
+	// Specifies the gzip compression level to use for the artifact.
```
Comment (simster7) [inline pkg/apis/workflow/v1alpha1/workflow_types.go:646]: Minor, but it's convention (at least in Argo) that docsrings start with the field name:
```suggestion
	// CompressionLevel specifies the gzip compression level to use for the artifact.
```

File: workflow/executor/docker/docker.go [hunk 2]
```diff
@@ -45,9 +46,24 @@ func (d *DockerExecutor) GetFileContents(containerID string, sourcePath string)
...
+	case gzip.DefaultCompression:
+		// use cmd default
+		levelFlag = ""
+	default:
+		// -1 through -9 (or error)
+		levelFlag = "-" + strconv.Itoa(compressionLevel)
+	}
+	if compressionLevel != gzip.DefaultCompression {
+
+	}
```
Comment (simster7) [inline workflow/executor/docker/docker.go:65]: Empty `if` block?

Comment (adamglt) [reply workflow/executor/docker/docker.go]: My bad - missed it.

**Comment by adamglt** [comment]
Not really sure where to add tests - couldn't find anything testing the exported archive (it seems functional tests run on the examples?).

Added to the archive example and some in the README.md.

File: util/archive/archive_test.go [hunk 3]
```diff
@@ -56,7 +57,7 @@ func TestTarFile(t *testing.T) {
 	log.Infof("Taring to %s", f.Name())
 	w := bufio.NewWriter(f)
 
-	err = TarGzToWriter(data.Name(), w)
+	err = TarGzToWriter(data.Name(), gzip.DefaultCompression, w)
```
Comment (simster7) [inline util/archive/archive_test.go:60]: For a test I think it would be enough to duplicate this with a different compression

**Comment by adamglt** [comment]
@simster7 is this planned for `2.7.4` or `2.8.x`?

**Comment by simster7** [comment]
@adamglt There should also be a `2.8-rc1` release in a week or so

**Comment by simster7** [comment]
@adamglt [`v2.8.0-rc1`](https://github.com/argoproj/argo/releases/tag/v2.8.0-rc1) has been released with this

**Pull request merged**